File: tests/oracle/printc_blockgraph_1204.cc
#include "printc.hh"
#include "block.hh"

#include <iostream>
#include <sstream>
#include <string>
#include <vector>

namespace {

using namespace ghidra;

struct BlockAddTag {
  using type = void (BlockGraph::*)(FlowBlock *);
  friend type access(BlockAddTag);
};

template <typename Tag, typename Tag::type Member>
struct PrivateAccess {
  friend typename Tag::type access(Tag) { return Member; }
};

template struct PrivateAccess<BlockAddTag, &BlockGraph::addBlock>;

class FixturePrintC;

class ProbeBlock final : public FlowBlock {
  int4 marker;
  block_type type;

public:
  ProbeBlock(int4 marker_in, block_type type_in)
      : marker(marker_in), type(type_in) {}

  virtual block_type getType(void) const { return type; }
  virtual void emit(PrintLanguage *language) const;
};

class FixturePrintC final : public PrintC {
  std::vector<int4> visits;

public:
  FixturePrintC(void) : PrintC(nullptr, "printc-blockgraph-1204") {
    delete emit;
    emit = new EmitNoMarkup();
  }

  void record(int4 marker) { visits.push_back(marker); }

  const std::vector<int4> &render(BlockGraph *graph) {
    emitBlockGraph(graph);
    return visits;
  }
};

void ProbeBlock::emit(PrintLanguage *language) const {
  FixturePrintC *printer = dynamic_cast<FixturePrintC *>(language);
  if (printer == nullptr)
    throw LowlevelError("unexpected fixture print language");
  printer->record(marker);
}

void append(BlockGraph &graph, FlowBlock *block) {
  const auto add = access(BlockAddTag{});
  (graph.*add)(block);
}

} // namespace

int main(void) {
  BlockGraph graph;
  append(graph, new ProbeBlock(17, FlowBlock::t_basic));
  append(graph, new ProbeBlock(3, FlowBlock::t_dowhile));
  append(graph, new ProbeBlock(29, FlowBlock::t_basic));

  FixturePrintC printer;
  const std::vector<int4> &visits = printer.render(&graph);
  std::cout << "visit_order=";
  for (size_t i = 0; i < visits.size(); ++i) {
    if (i != 0)
      std::cout << ',';
    std::cout << visits[i];
  }
  std::cout << '\n';
  std::cout << "visit_count=" << visits.size() << '\n';
  std::cout << "dowhile_visits=";
  size_t loop_count = 0;
  for (const int4 marker : visits) {
    if (marker == 3)
      ++loop_count;
  }
  std::cout << loop_count << '\n';
  return 0;
}

File: tests/oracle/printc_blockgraph_1204.rs
use rugra::address::{Address, SeqNum};
use rugra::block::{BlockBasic, BlockDoWhile, BlockGraph, FlowBlock};
use rugra::funcdata::Funcdata;
use rugra::op::{PcodeOp, PcodeOpRef};
use rugra::opcodes::OpCode;
use rugra::prettyprint::EmitNoMarkup;
use rugra::printc::PrintC;
use rugra::space::AddressSpace;
use rugra::varnode::Varnode;
use std::sync::{Arc, RwLock};

type DynBlock = Arc<RwLock<dyn FlowBlock + Send + Sync>>;

fn make_op(opcode: OpCode, index: u32) -> PcodeOp {
    let mut op = PcodeOp::new(SeqNum::new(Address::new(index as u64), index), opcode);
    op.set_opcode_flags(opcode);
    op
}

fn make_call_block(target: u64, index: i32) -> DynBlock {
    let mut op = make_op(OpCode::CPUI_CALL, index as u32);
    op.inrefs.push(Arc::new(RwLock::new(Varnode::new_with_space(
        8,
        AddressSpace::Const,
        target,
    ))));
    let mut block = BlockBasic::new(index, Address::new(index as u64));
    block.add_op(PcodeOpRef(Arc::new(RwLock::new(op))));
    Arc::new(RwLock::new(block))
}

fn graph_visit_order() -> Vec<u64> {
    let targets = [17_u64, 3, 29];
    let mut graph = BlockGraph::new();
    for (index, target) in targets.iter().copied().enumerate() {
        graph.add_block(make_call_block(target, index as i32));
    }

    let mut printer = PrintC::new(Box::new(EmitNoMarkup::new()));
    printer.emit_block_graph(&graph);
    let emit = printer
        .take_emit()
        .into_any()
        .downcast::<EmitNoMarkup>()
        .expect("PrintC block-graph fixture must retain EmitNoMarkup");
    let output = emit.debug_get_output_ref();

    let mut positions = targets
        .iter()
        .copied()
        .map(|target| {
            let name = format!("FUN_{target:x}");
            let position = output
                .find(&name)
                .unwrap_or_else(|| panic!("missing {name} in {output:?}"));
            (position, target)
        })
        .collect::<Vec<_>>();
    positions.sort_by_key(|(position, _)| *position);
    positions.into_iter().map(|(_, target)| target).collect()
}

fn make_do_while() -> DynBlock {
    let mut branch = make_op(OpCode::CPUI_CBRANCH, 1);
    branch
        .inrefs
        .push(Arc::new(RwLock::new(Varnode::new_with_space(
            8,
            AddressSpace::Const,
            0x1000,
        ))));
    branch
        .inrefs
        .push(Arc::new(RwLock::new(Varnode::new_with_space(
            1,
            AddressSpace::Const,
            1,
        ))));
    let mut condition = BlockBasic::new(0, Address::new(0x1000));
    condition.add_op(PcodeOpRef(Arc::new(RwLock::new(branch))));
    let condition: DynBlock = Arc::new(RwLock::new(condition));
    Arc::new(RwLock::new(BlockDoWhile {
        index: 0,
        condition,
        incoming: Vec::new(),
        outgoing: Vec::new(),
        parent: None,
        flags: 0,
    }))
}

fn doc_function_dowhile_visits() -> usize {
    let mut function = Funcdata::new("blockgraph_probe", Address::new(0x1000), 1);
    function.sblocks.add_block(make_do_while());

    let mut printer = PrintC::new(Box::new(EmitNoMarkup::new()));
    printer.doc_function_inherent(&function);
    let emit = printer
        .take_emit()
        .into_any()
        .downcast::<EmitNoMarkup>()
        .expect("PrintC doc-function fixture must retain EmitNoMarkup");
    emit.debug_get_output_ref().matches("do ").count()
}

fn main() {
    let visits = graph_visit_order();
    println!(
        "visit_order={}",
        visits
            .iter()
            .map(u64::to_string)
            .collect::<Vec<_>>()
            .join(",")
    );
    println!("visit_count={}", visits.len());
    println!("dowhile_visits={}", doc_function_dowhile_visits());
}
